server/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsServiceImpl.java

```diff
@@ -126,10 +185,336 @@ protected String prepareShellCmd(String cmdType, String ipAddress, String option
+        }
+        return fileList;
+    }
+
+    private Pair<Boolean, String> orchestrateCopyToSecondaryStorageNonVMware(final DataStore store, final String vmControlIp, String fileToCopy, Long vmHostId) {
+        CopyToSecondaryStorageCommand toSecondaryStorageCommand = new CopyToSecondaryStorageCommand(store.getUri(), vmControlIp, fileToCopy);
+        Answer copyToSecondaryAnswer = agentManager.easySend(vmHostId, toSecondaryStorageCommand);
+        Pair<Boolean, String> copyAnswer;
+        if (copyToSecondaryAnswer != null) {
+            copyAnswer = new Pair<>(copyToSecondaryAnswer.getResult(), copyToSecondaryAnswer.getDetails());
+        } else {
+            copyAnswer = new Pair<>(false, "Diagnostics Zip file to secondary storage failed");
+        }
+        return copyAnswer;
+    }
+
+    private Pair<Boolean, String> orchestrateCopyToSecondaryStorageVMware(final DataStore store, final String vmSshIp, String diagnosticsFile) {
+        String mountPoint;
+        boolean success;
+
+        Integer nfsVersion = imageStoreDetailsUtil.getNfsVersion(store.getId());
+        mountPoint = mountManager.getMountPoint(store.getUri(), nfsVersion);
+        if (StringUtils.isNotBlank(mountPoint)) {
+            LOGGER.info(String.format("Copying %s from %s to secondary store %s", diagnosticsFile, vmSshIp, store.getUri()));
+
+            // dirIn/mnt/SecStorage/uuid/diagnostics_data
+            String dataDirectoryInSecondaryStore = String.format("%s/%s", mountPoint, DIAGNOSTICS_DIRECTORY);
+            try {
+                File dataDirectory = new File(dataDirectoryInSecondaryStore);
+                boolean existsInSecondaryStore = dataDirectory.exists() || dataDirectory.mkdir();
+
+                if (existsInSecondaryStore) {
+                    // scp from system VM to mounted sec storage directory
+                    int port = 3922;
+                    File permKey = new File("/var/cloudstack/management/.ssh/id_rsa");
+                    SshHelper.scpFrom(vmSshIp, port, "root", permKey, dataDirectoryInSecondaryStore, diagnosticsFile);
+                }
+                // Verify File copy to Secondary Storage
+                File fileInSecondaryStore = new File(dataDirectoryInSecondaryStore + diagnosticsFile.replace("/root", ""));
+                success = fileInSecondaryStore.exists();
+            } catch (Exception e) {
+                String msg = String.format("Exception caught during scp from %s to secondary store %s: ", vmSshIp, dataDirectoryInSecondaryStore);
+                LOGGER.error(msg, e);
+                return new Pair<>(false, msg);
+            } finally {
+                // umount secondary storage
+                umountSecondaryStorage(mountPoint);
+            }
+        } else {
+            return new Pair<>(false, "Failed to mount secondary storage:" + store.getName());
+        }
+        return new Pair<>(success, "File copied to secondary storage successfully");
+    }
```

Comment:
same comment as above + can you partition this method?

Reply:
Done